byterun/memory.c

```diff
@@ -574,29 +576,157 @@ CAMLexport CAMLweakdef void caml_modify (value *fp, value val)
   }
 }
 
+
+/* Global memory pool.
+
+   The pool is structured as a ring of blocks, where each block's header
+   contains two links: to the previous and to the next block. The data structure
+   allows performing insertions and removals of blocks in constant time, given
+   that a pointer to the block is provided.
+
+   Initially, the pool contains a single block -- a pivot with no data, the
+   guaranteed existence of which makes for a more concise implementation.
+
+   The API functions that operate on the pool receive not pointers to the
+   block's header, but rather pointers to the block's "data" field. This
+   behavoir is required to maintain compatibility with the interfaces of
+   malloc, realloc, free, as well as to hide the implementation from the user.
+*/
+
+// A type with the most strict alignment requirements
+union max_align {
+  char c;
+  short s;
+  long l;
+  int i;
+  float f;
+  double d;
+  void *v;
+  void (*q)(void);
+};
+
+struct pool_block {
+#ifdef DEBUG
+  long magic;
+#endif
+  struct pool_block *next;
+  struct pool_block *prev;
+  union max_align data[1];  /* not allocated, used for alignment purposes */
+};
+
+#define SIZEOF_POOL_BLOCK offsetof(struct pool_block, data)
+
+static struct pool_block *pool = NULL;
+
+
+/* Returns a pointer to the block header, given a pointer to "data" */
+static struct pool_block* get_pool_block(void *ptr)
+{
+    if (ptr == NULL) return NULL;
+    struct pool_block *pb = (void*)(ptr) - SIZEOF_POOL_BLOCK;
```

Comment:
This should be `(void *) ((char*) ptr - SIZEOF_POOL_BLOCK)` because you can't do arithmetic on a `void *`.